lib/active_job/limiter/mixin.rb

```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
             ActiveJob::Limiter.clear_lock_before_perform(job)
           end
         end
+
+        def throttle_job(duration:, extract_resource_id:, metrics_hook: ->(_result, _job) {})
+          active_job_limiter_add_throttle_job_around_enqueue(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+          active_job_limiter_add_throttle_job_around_perform(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+        end
+
+        private
+
+        # Prefixing private methods with active_job_limiter to avoid conflicts with user code
+
+        def active_job_limiter_add_throttle_job_around_enqueue(duration:, extract_resource_id:, metrics_hook:)
+          around_enqueue do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if job.instance_variable_get(:@bypass_active_job_limiter_enqueue_locks)
+              block.call
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('enqueue', duration, job, resource_id)
```

Comment:
possible that we acquire lock but crash before enqueue operation...

Reply:
In that case Sidekiq should retry the job, right? Which should re-run the around_perform hook, and lead to a reschedule to occur (since the perform lock will have been acquired).

Of course, the server could also crash between getting the reschedule lock and scheduling the job for the future.

That would be a once-in-many-billion event, though. To me, that is not worth diving into what I’m sure would be activejob/sidekiq internals to try to atomically lock-and-schedule. limit_queue has been serving us well for years without this feature.